ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OAlgorithm.scala

```diff
@@ -37,31 +42,56 @@ import scala.reflect.{ClassTag, classTag}
  * Base class for H2O algorithm wrapper as a Spark transformer.
  */
 abstract class H2OAlgorithm[B <: H2OBaseModelBuilder : ClassTag, M <: H2OBaseModel, P <: Model.Parameters : ClassTag]
-  extends Estimator[H2OMOJOModel] with H2OAlgoCommonUtils with DefaultParamsWritable with H2OAlgoCommonParams[P] {
+  extends Estimator[H2OMOJOModel]
+    with H2OAlgoCommonUtils
+    with DefaultParamsWritable
+    with H2OAlgoCommonParams[P]
+    with RestCommunication {
 
-  protected def preProcessBeforeFit(trainFrameKey: String): Unit = {}
+  protected def prepareH2OTrainFrameForFitting(trainFrameKey: String): Unit = {}
 
-  override def fit(dataset: Dataset[_]): H2OMOJOModel = {
-    // Update H2O params based on provided configuration
-    updateH2OParams()
+  private def fitOverRest(trainKey: String, validKey: Option[String]): Array[Byte] = {
+    prepareH2OTrainFrameForFitting(trainKey)
+    val conf = H2OContext.ensure().getConf
+    val params = getH2OAlgoRESTParams() ++
+      Map("training_frame" -> trainKey) ++
+      validKey.map { key => Map("validation_frame" -> key) }.getOrElse(Map())
 
-    val (trainKey, validKey, internalFeatureCols) = prepareDatasetForFitting(dataset)
+    val algoName = parameters.algoName().toLowerCase
+    val endpoint = RestApiUtils.getClusterEndpoint(conf)
+    val content = withResource(readURLContent(endpoint, "POST", s"/3/ModelBuilders/$algoName", conf, params)) { response =>
+      IOUtils.toString(response)
+    }
+
+    val gson = new Gson()
+    val job = gson.fromJson(content, classOf[JsonElement]).getAsJsonObject.get("job").getAsJsonObject
+    val jobId = job.get("key").getAsJsonObject.get("name").getAsString
+    H2OJob(jobId).waitForFinish()
+    val modelId = job.get("dest").getAsJsonObject.get("name").getAsString
+    H2OModel(modelId).getOrDownloadMojoData()
+  }
+
+  private def fitOverClient(trainKey: String, validKey: Option[String]): Array[Byte] = {
+    updateH2OParams()
     parameters._train = DKV.getGet[Frame](trainKey)._key
     parameters._valid = validKey.map(DKV.getGet[Frame](_)._key).orNull
+    prepareH2OTrainFrameForFitting(trainKey)
+    water.DKV.put(parameters._train.get())
+    val binaryModel = trainModel(parameters)
+    ModelSerializationSupport.getMojoData(binaryModel)
+  }
 
-    val trainFrame = parameters._train.get()
-    preProcessBeforeFit(trainFrame._key.toString)
-    if (!RestApiUtils.isRestAPIBased()) {
-      water.DKV.put(trainFrame)
+  override def fit(dataset: Dataset[_]): H2OMOJOModel = {
+    val (trainKey, validKey, internalFeatureCols) = prepareDatasetForFitting(dataset)
+    val mojoData = if (RestApiUtils.isRestAPIBased()) {
```

Comment:
Are Algo tests in scala executed in both modes?

Reply:
We have only training over rest now, so all tests are running like that

Reply:
ok, SGTM